WordPress/Classes/Utility/Media/UIImageView+ImageDownloader.swift

```diff
@@ -0,0 +1,57 @@
+import Foundation
+import UIKit
+import Gifu
+
+extension UIImageView {
+    @MainActor
+    var wp: ImageViewExtensions { ImageViewExtensions(imageView: self) }
+}
+
+@MainActor
+struct ImageViewExtensions {
+    var imageView: UIImageView
+
+    func prepareForReuse() {
+        controller.prepareForReuse()
+
+        if let gifView = imageView as? GIFImageView, gifView.isAnimatingGIF {
+            gifView.prepareForReuse()
+        } else {
+            imageView.image = nil
+        }
+    }
+
+    func setImage(with imageURL: URL, host: MediaHost? = nil, size: CGSize? = nil) {
+        controller.setImage(with: imageURL, host: host, size: size)
+    }
+
+    var controller: ImageViewController {
+        if let controller = objc_getAssociatedObject(imageView, ImageViewExtensions.controllerKey) as? ImageViewController {
+            return controller
+        }
+        let controller = ImageViewController()
+        controller.onStateChanged = { [weak imageView] in
+            guard let imageView else { return }
+            setState($0, for: imageView)
+        }
+        objc_setAssociatedObject(imageView, ImageViewExtensions.controllerKey, controller, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
+        return controller
+    }
+
+    private func setState(_ state: ImageViewController.State, for imageView: UIImageView) {
+        switch state {
+        case .loading:
+            break
+        case .success(let image):
+            if let gifView = imageView as? GIFImageView {
+                gifView.configure(image: image)
+            } else {
+                imageView.image = image
+            }
+        case .failure:
+            break
+        }
+    }
+
+    private static let controllerKey = malloc(1)!
```

Comment:
This is the first time I see this pattern. The pattern used in most of the code is using the address of a static variable. Like this code:

```swift
private var observerKey = 0

extension StatsForegroundObservable where Self: UIViewController {
    private var enterForegroundObserver: NSObjectProtocol? {
        get {
            objc_getAssociatedObject(self, &controllerKey) as? NSObjectProtocol
        }
        set {
            objc_setAssociatedObject(self, &observerKey, newValue, .OBJC_ASSOCIATION_RETAIN)
        }
    }
}
```

Reply:
It's one of the modern, if that's that the right word, ways to do it. I originally learned about it from Eskimo's [message [here](https://forums.swift.org/t/objc-setassociatedobject-vs-swift/45188/3), and since then shipped it [here](https://github.com/kean/Nuke/blob/0e953da39106db3f989364ba443a8ca12e558668/Sources/NukeExtensions/ImageViewExtensions.swift#L221), so it's been battle tested.

This will produce a warning.

```swift
objc_getAssociatedObject(self, &controllerKey)
```

Reply:
I don't think it makes much sense to use `&controllerKey`, given `controllerKey` is already a pointer.

To be honest, both approaches seem a bit strange to me, even though they all make perfect sense.